# Reading Data From Twitter

Install tweepy

In [1]:
!pip install git+https://github.com/tweepy/tweepy.git

Defaulting to user installation because normal site-packages is not writeable


In [ ]:
!pip install TextBlob

## Insert bearer token

1. You will need a bearer token from the twitter development portal: https://developer.twitter.com/en
2. Place the token in a variable

In [7]:
bearer_token = 'AAAAAAAAAAAAAAAAAAAAAD7cjQEAAAAA30Ipcn3vDlWpr6XIj0h0xZDa4io%3DVL20hy1g6ZsgLnvZOxMX24F7w3IaHRzTIB1CibUpu5XmZsZqXx'

Create a client with your bearer_token

In [8]:
import tweepy as tw
client = tw.Client(bearer_token=bearer_token)

Search recent tweets from the official HSG Twitter account @HSGStGallen

In [9]:
response = client.search_recent_tweets('(from:HSGStGallen)')
response

Response(data=[<Tweet id=1597865509005176832 text='Gerade bei #Krisen wie der #Pandemie ist wissenschaftliche Expertise wichtig. In politischem Handeln münden die Fakten aber oft nicht. Warum ist das Verhältnis #Wissenschaft &amp; #Politik so schwierig? #HSGProf Caspar Hirschi @hsg_shss war zu Gast bei @3sat https://t.co/1gwkJ3A42f https://t.co/CeKPCZi0ra'>, <Tweet id=1597501566818582532 text='Morgen ist Andreas Hänni zu Gast im Sports Economics Talk. Er ist ein ehemaliger Schweizer Eishockeyprofi und im Jahr 2016 gründete er die Firma «49 mining, analytics &amp; consulting GmbH», zu deren Kunden alle Schweizer Profi-Eishockeymannschaften und einige Amateurvereine zählen. https://t.co/xzLOelviK0'>, <Tweet id=1597207834005233664 text='Die HSG bildet Medizinstudierende an ihrer School of Medicine für den Ernstfall aus: Die angehenden Ärztinnen und Ärzte trainieren bei der Behandlung von Schauspieler:innen, die medizinische Probleme vorspielen, für die Berufspraxis.\n\nhttps://t.co/74k1mu

We get an response object with
- data: The returned tweets
- meta: A dictionary with additional information. We got the information that 10 results have been returned along with a next_token

Try to request more tweets from HSG. The documentation tells us that this endpoint allows a maximum number of tweets per request of 100 and returns the tweets of the last seven days

In [10]:
response = client.search_recent_tweets('(from:HSGStGallen)', max_results=100)
response

Unauthorized: 401 Unauthorized
Unauthorized

Around 10 tweets, not too many. Try to request all tweets from HSG

In [54]:
response = client.search_all_tweets('(from:HSGStGallen)', max_results=100)
response

Forbidden: 403 Forbidden
When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal.

The API returns a '403 Forbidden' status code. It seems we have correct authentication (otherwise it would be 401 Unauthenticated), but are missing the required access rights. The documentation informs us, that the full-archive search requires academic research access. 

# Pagination

Search for more tweets: Get all recent tweets that contain the hashtag FIFAWORLDCUP and are written in German

In [4]:
query = "#FifaWorldCup lang:en"
response = client.search_recent_tweets(query, max_results=100)
response

Response(data=[<Tweet id=1596151603358113794 text='RT @MimosaFawaz: Senegal 🇸🇳 fans in fine voice supporting the Teranga Lions 🦁 in the stadium. \n\n#FIFAWorldCup #FIFAWorldCupQatar2022 \n\n🎥 @P…'>, <Tweet id=1596151602074726406 text='RT @premierleague: A #FIFAWorldCup brace for Richarlison 🔥\n\n@SpursOfficial | @richarlison97 https://t.co/N1eQdxfaNf'>, <Tweet id=1596151601139699713 text='RT @FiercePC: ⚽️ netherlands vs ecuador⚽️\n\nif netherlands score EXACTLY 3 GOALS vs ecuador we’ll pick a winner and send a custom gaming pc…'>, <Tweet id=1596151600850034689 text='RT @FIFAWorldCup: A huge win for IR Iran! 👏\n\n#FIFAWorldCup | #Qatar2022'>, <Tweet id=1596151600795508736 text='RT @JoueursSN: Le but de Famara Diedhiou en video ! 🎥\n\n#QATSEN | #FIFAWorldCup \n\n https://t.co/sqx0yweCTK'>, <Tweet id=1596151600229359616 text='@focusonJK97 @BTS_twt I’m listening to #Jungkook of @BTS_twt  new single "Dreamers" for the #FIFAWorldCup     #Dreamers2022'>, <Tweet id=1596151599394701312 text='If

Thats a lot! It seems we have hit the maximum number of results per request

In [5]:
print(response.meta["result_count"])
print(len(response.data))

100
100


The API provides aus with a next_token meta information, to indicate that there are more results. But as only 100 tweets are returned per request, we need to start a new request. 

In [6]:
next_token = response.meta["next_token"]
next_token

'b26v89c19zqg8o3fpzhlqmd8a7985914sej8l5559jbzx'

Request the next 100 tweets for this search

In [7]:
response = client.search_recent_tweets(query, max_results=100, next_token=next_token)
response

Response(data=[<Tweet id=1596151531107389440 text='RT @JKTEAM_KOR: 🌟STREAM FOR JUNGKOOK \n\nOur goal is 1000screenshots \nPlease drop your stream with the following tags.\n\nI am so amazed by #Ju…'>, <Tweet id=1596151528280457218 text='Substitute Rouzbeh Cheshmi and right-back Ramin Rezaeian scored the goals in their FIFA World Cup Group B clash at Ahmad Bin Ali Stadium in Al Rayyan on Friday\n\n#FIFAWorldCup #Qatar2022\n\nBy @vivek9301\n\nhttps://t.co/wJRo9GgNMp'>, <Tweet id=1596151527034744834 text="RT @SofascoreINT: ⚠️ | QUICK STAT\n\nMehdi Taremi's 8 key passes are the most by a player in one #FIFAWorldCup match since Kevin De Bruyne's…">, <Tweet id=1596151527034744833 text='RT @Daily_JK97: 📢 PH ARMYS, WE WILL BE FUNDING “DREAMERS” by JK\n\n♦️How to request fund, send me: \n\n✔Gcash no. / Smart No. \n✔ Send Proof of…'>, <Tweet id=1596151526791073793 text='RT @alimo_philip: 🏟️ @ASAMOAH_GYAN3 &amp; Cristiano Ronaldo are the only players to score in 9 &amp; 10 major international tour

We got another 100 tweets. This process is called *Pagination*. Imagine a telefone book: Each page can store 100 tweets and you have to turn to the next page to get more information.

To ease this process tweepy has a Paginator object 

In [11]:
tweets_paginator = tw.Paginator(client.search_recent_tweets, query, max_results=100).flatten(limit=250)
tweets_paginator

<generator object Paginator.flatten at 0x000001B59B0A9DD0>

We got a paginator object that will handle the requests for us. When we iterate over the object, it will return the first 250 tweets. 

In [10]:
for tweet in tweets_paginator:
    print(tweet.text)

Generator are one time use objects. Rerunning the below code cell results in no output

In [121]:
for tweet in tweets_paginator:
    print(tweet.text)

To store the tweets in a persistant manner, we are storring them in a list. Remember to rerun the Pagination request each time!

In [122]:
tweets_paginator = tw.Paginator(client.search_recent_tweets, query, max_results=100).flatten(limit=250)

tweets = []

for tweet in tweets_paginator:
    tweets.append(tweet.text)

tweets

['RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttps://t.co/l841e5rbj7\n\n@SolidarityUKR #Ukraine #…',
 "RT @jcbehrends: Personally, I don't believe #Kadyrov or #Prigozhin will topple #Putin. \nI regard them as warlords. They stand for the disin…",
 'RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttps://t.co/l841e5rbj7\n\n@SolidarityUKR #Ukraine #…',
 'RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttps://t.co/l841e5rbj7\n\n@SolidarityUKR #Ukraine #…',
 'RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttps://t.co/l841e5rbj7\n\n@SolidarityUKR #Ukraine #…',
 'RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttps://t.co/l841e5rbj7\n\n@SolidarityUKR #Ukraine #…',
 'RT @pressat: RED BULL HEADQUARTERS HIT BY UKRAINE PROTEST\n\n"RED BULL GIVES PUTIN WINGS"\n\nhttp

# Sentiment Analysis Tweets


In [123]:
import pandas as pd
tweets_df = pd.DataFrame(data=tweets, columns=['Tweets']).drop_duplicates()
tweets_df

Tweets
0    RT @pressat: RED BULL HEADQUARTERS HIT BY UKRA...
1    RT @jcbehrends: Personally, I don't believe #K...
8    @GlasnostGone @ZelenskyyUa A true leader unlik...
14   #MikePompeo proved he’s not ready for Prime ti...
19   RT @usownstheplanet: The Embassy would like to...
22   #Ukraine after russia’s latest attack on our c...
26   If you don’t own ANY $FBX you are NOT prepared...
27   @SenRonJohnson @SenateGOP yeah, the Russian me...
31   RT @wilkoskillz: Good morning sir , \nYour att...
32   They never learnt their lesson!! 🔥💪👑❤️‍🔥 #Trut...
41   RT @ltgrusselhonore: Let’s Be Clear #Putin and...
43   RT @georgegalloway: The one and only Ex CIA ma...
49   If I Had A Tiger On Hand? #Putin   🐅 https://t...
58   RT @agwu_isaiah: @ChinasaNworu @mfa_russia @Ru...
61   RT @ChinasaNworu: #Putin Sir , Your attention ...
64   #Putin #Maduro, #noamchomsky himself agrees US...
74   200 kg of seized cannabis was eaten by rats, s...
79   RT @MoistureVapor8r: @BlackKnight10k #Manafort...
80   RT @EmbassyofRussia: President Vladimir #Putin...
81   Also while #FIFAWorldCup, we should never forg...
82   No, there's no Nazism there!! #Truth #Justice ...
83   @elonmusk https://t.co/EuFqwwLSav The Exposing...
90   RT @AhmadSuriy: #Humiliation\n\nThe man asks t...
91   @blueheartedly YES as #Putin is still holding ...
96   #Lukashenko acted as a #negotiator, where he p...
111  RT @toryboypierce: If @BorisJohnson right and ...
112  @theurbanmammal @IuliiaMendel That settles it....
113  RT @jay4_nsa: Someone told Russia's Tucker Car...
114  Since when the wish for peace is considered "d...
117  #BREAKING\nMORE reporting that #Russia intends...
142     @lookonchain @VitalikButerin #Putin needs ammo
146  The Duma needs to pick up another issue more R...
151  RT @Iykayofficial: The former Peugeot factory ...
154  Most everyone knows how to quickly end russia’...
162  RT @MCTothSTL: @KyivPost @EliotHiggins #Putin’...
163  My latest w/@JESweet2022 at @thehill / @TheHil...
170  RT @ignis_fatum: ⚠️#putin regime terrorists an...
171  RT @cartoonmovement: Russian men are attemptin...
172  #Ukraine #Putin reportedly very afraid for his...
179  CSTO has collapsed???\n\n#Putin #Ukraine #Russ...
183                             @Reuters Thanks #Putin
184  EU Commission called on the EU countries to se...
186  " #PUTIN: \n#OIL #PRICE #CAP ON #RUSSIA WILL H...
187  RT @InnaSovsun: We want to see the recognition...
189  I am very̡ happy with my website and its conte...
192  #Ukraine braces for sub-zero temperatures as p...
194  Meghan Markle has turned into ‘sad desperate p...
199  RT @ipobTamale: #Putin Sir , Your attention is...
207  The Metro of Kiev!! And despite that this time...
209  RT @KarlibayBEKMUR2: @OSW_pl #Putin's puppets ...
211  RT @kiraincongress: #Putin thinks that one or ...
218  RT @EvilDave_NXT: @Sputnik_Not Imagine what Ru...
228  RT @Aigul_Nurpeis: 9/10\nWhat role is #Putin n...
230  RT @Iykayofficial: This isn't Mbano, #Biafra #...
232  #Putin awards Order of Courage to war correspo...
233  Amidst an unsuccessful war effort and a botche...
239  If @BorisJohnson right and #germany wanted qui...
246  While United Nations stands and watches the id...

Recap Module 2: text mining and sentiment analysis with TextBlob

In [124]:
from textblob import TextBlob

# function that only returns the polarity score of TextBlob
def polarity(text):
    try:
        return TextBlob(text).sentiment.polarity
    except:
        return None

# function that only returns the subjectivity score of TextBlob
def subjectivity(text):
    try:
        return TextBlob(text).sentiment.subjectivity
    except:
        return None

In [125]:
tweets_df["Polarity"]= tweets_df["Tweets"].apply(polarity)
tweets_df["Subjectivity"]= tweets_df["Tweets"].apply(subjectivity)
tweets_df

Tweets  Polarity  Subjectivity
0    RT @pressat: RED BULL HEADQUARTERS HIT BY UKRA...  0.000000      0.000000
1    RT @jcbehrends: Personally, I don't believe #K...  0.000000      0.300000
8    @GlasnostGone @ZelenskyyUa A true leader unlik...  0.112500      0.512500
14   #MikePompeo proved he’s not ready for Prime ti... -0.109524      0.577381
19   RT @usownstheplanet: The Embassy would like to...  0.000000      0.000000
22   #Ukraine after russia’s latest attack on our c...  0.250000      0.850000
26   If you don’t own ANY $FBX you are NOT prepared...  0.300000      0.562500
27   @SenRonJohnson @SenateGOP yeah, the Russian me...  0.125000      0.100000
31   RT @wilkoskillz: Good morning sir , \nYour att...  0.700000      0.600000
32   They never learnt their lesson!! 🔥💪👑❤️‍🔥 #Trut...  0.000000      0.000000
41   RT @ltgrusselhonore: Let’s Be Clear #Putin and... -0.200000      0.691667
43   RT @georgegalloway: The one and only Ex CIA ma...  0.000000      1.000000
49   If I Had A Tiger On Hand? #Putin   🐅 https://t...  0.000000      0.000000
58   RT @agwu_isaiah: @ChinasaNworu @mfa_russia @Ru...  0.000000      0.000000
61   RT @ChinasaNworu: #Putin Sir , Your attention ...  0.160000      0.540000
64   #Putin #Maduro, #noamchomsky himself agrees US...  0.000000      0.000000
74   200 kg of seized cannabis was eaten by rats, s...  0.000000      0.000000
79   RT @MoistureVapor8r: @BlackKnight10k #Manafort...  0.000000      0.000000
80   RT @EmbassyofRussia: President Vladimir #Putin... -0.012500      0.287500
81   Also while #FIFAWorldCup, we should never forg...  0.000000      0.500000
82   No, there's no Nazism there!! #Truth #Justice ...  0.000000      0.000000
83   @elonmusk https://t.co/EuFqwwLSav The Exposing...  0.000000      0.000000
90   RT @AhmadSuriy: #Humiliation\n\nThe man asks t...  0.000000      0.000000
91   @blueheartedly YES as #Putin is still holding ...  0.000000      0.000000
96   #Lukashenko acted as a #negotiator, where he p...  0.050000      0.300000
111  RT @toryboypierce: If @BorisJohnson right and ...  0.473016      0.478571
112  @theurbanmammal @IuliiaMendel That settles it....  0.000000      0.875000
113  RT @jay4_nsa: Someone told Russia's Tucker Car...  0.000000      0.000000
114  Since when the wish for peace is considered "d... -0.600000      0.900000
117  #BREAKING\nMORE reporting that #Russia intends...  0.050000      0.375000
142     @lookonchain @VitalikButerin #Putin needs ammo  0.000000      0.000000
146  The Duma needs to pick up another issue more R...  0.100505      0.422222
151  RT @Iykayofficial: The former Peugeot factory ...  0.500000      0.500000
154  Most everyone knows how to quickly end russia’...  0.244167      0.460000
162  RT @MCTothSTL: @KyivPost @EliotHiggins #Putin’...  0.100000      0.400000
163  My latest w/@JESweet2022 at @thehill / @TheHil... -0.037500      0.712500
170  RT @ignis_fatum: ⚠️#putin regime terrorists an... -0.100000      0.100000
171  RT @cartoonmovement: Russian men are attemptin... -0.050000      0.050000
172  #Ukraine #Putin reportedly very afraid for his... -0.780000      1.000000
179  CSTO has collapsed???\n\n#Putin #Ukraine #Russ...  0.000000      0.000000
183                             @Reuters Thanks #Putin  0.200000      0.200000
184  EU Commission called on the EU countries to se...  0.000000      0.000000
186  " #PUTIN: \n#OIL #PRICE #CAP ON #RUSSIA WILL H... -0.027778      0.333333
187  RT @InnaSovsun: We want to see the recognition...  0.000000      0.000000
189  I am very̡ happy with my website and its conte...  0.800000      1.000000
192  #Ukraine braces for sub-zero temperatures as p...  0.000000      0.750000
194  Meghan Markle has turned into ‘sad desperate p... -0.550000      1.000000
199  RT @ipobTamale: #Putin Sir , Your attention is...  0.160000      0.540000
207  The Metro of Kiev!! And despite that this time...  0.390625      0.300000
209  RT @KarlibayBEKMUR2: @OSW_pl #Putin's puppets ...  0.000000      0.500000
211  RT @kiraincong

In [126]:
tweets_df.mean()

C:\Users\IBlohm\AppData\Local\Temp\ipykernel_8536\1926658668.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  tweets_df.mean()


Polarity        0.042639
Subjectivity    0.342943
dtype: float64